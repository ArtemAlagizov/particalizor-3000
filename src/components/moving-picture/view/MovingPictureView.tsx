import React from "react";
import { Styled } from "../styled";
import { IMovingPictureViewProps } from "./IMovingPictureViewProps";

const MovingPictureView: React.FC<IMovingPictureViewProps> = ({
  destinationCanvasRef,
  referenceCanvasRef
}) => {
  return (
    <>
      <Styled.DestinationCanvas>
        <canvas ref={destinationCanvasRef} className="dest-canvas" />
      </Styled.DestinationCanvas>
      <Styled.RefCanvas>
        <canvas id="referenceCanvas" ref={referenceCanvasRef} />
      </Styled.RefCanvas>
    </>
  );
};

export default MovingPictureView;
