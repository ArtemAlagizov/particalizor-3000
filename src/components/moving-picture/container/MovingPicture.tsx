import React, { useEffect, useRef } from "react";
import MovingPictureView from "../view/MovingPictureView";
import { IMovingPictureProps } from "./IMovingPictureProps";
import { getActualValues } from "../../../services/properties-handler/PropertiesHandler";
import {
  clearCanvas,
  createImageElement
} from "../../../services/painter/Painter";

export const MovingPicture: React.FC<IMovingPictureProps> = ({
  particleNumber,
  particleLifeTime,
  particleVelocity,
  directionChannel,
  reverseDirection,
  hueChannel,
  reverseHue,
  particleTraceWidth,
  imageSource
}) => {
  const destinationCanvasRef = useRef<HTMLCanvasElement>(null);
  const referenceCanvasRef = useRef<HTMLCanvasElement>(null);

  const actualValues = getActualValues(
    particleNumber,
    particleLifeTime,
    particleVelocity,
    directionChannel,
    reverseDirection,
    hueChannel,
    reverseHue,
    particleTraceWidth
  );

  let animationRequestId: number | null = 0;

  const getAnimationRequestId = () => animationRequestId;
  const setAnimationRequestId = (requestId: number | null) =>
    (animationRequestId = requestId);

  useEffect(() =>
    createImageElement(
      referenceCanvasRef.current,
      destinationCanvasRef.current,
      actualValues,
      imageSource,
      setAnimationRequestId,
      getAnimationRequestId
    )
  );

  useEffect(() => {
    return () => {
      setAnimationRequestId(null);

      clearCanvas(destinationCanvasRef);
    };
  }, [destinationCanvasRef]);

  return (
    <MovingPictureView
      key={Math.random()}
      destinationCanvasRef={destinationCanvasRef}
      referenceCanvasRef={referenceCanvasRef}
    />
  );
};

export default MovingPicture;
