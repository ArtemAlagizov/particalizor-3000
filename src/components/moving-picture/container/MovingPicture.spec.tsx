import React from "react";
import { mount, shallow } from "enzyme";
import MovingPicture from "./MovingPicture";
import MovingPictureView from "../view/MovingPictureView";

describe("Given a MovingPicture", () => {
  describe("when rendering", () => {
    it("should match snapshot", () => {
      Math.random = jest.fn(() => 777);

      const component = shallow(<MovingPicture imageSource={""} />);

      expect(component).toMatchSnapshot();
    });
    it("should render MovingPictureView", () => {
      const component = mount(<MovingPicture imageSource={""} />);

      expect(component.find(MovingPictureView).length).toBe(1);
    });
  });
});
