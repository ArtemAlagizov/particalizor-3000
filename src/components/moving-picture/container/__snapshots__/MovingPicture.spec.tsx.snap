// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a MovingPicture when rendering should match snapshot 1`] = `
<MovingPictureView
  destinationCanvasRef={
    Object {
      "current": null,
    }
  }
  key="777"
  referenceCanvasRef={
    Object {
      "current": null,
    }
  }
/>
`;
